server/chat-plugins/repeats.ts

```diff
@@ -0,0 +1,198 @@
+/**
+ * Chat plugin for repeating messages in chat
+ * Based on bot functionality from Kid A and Expecto Botronum
+ * @author Annika, Zarel
+ */
+
+export interface RepeatedPhrase {
+	phrase: string;
+	/** interval in milliseconds */
+	interval: number;
+}
+
+export const Repeats = new class {
+	// keying to Room rather than RoomID will help us correctly handle room renames
+	/** room:phrase:timeout map */
+	repeats = new Map<BasicRoom, Map<string, NodeJS.Timeout>>();
+
+	constructor() {
+		for (const room of Rooms.rooms.values()) {
+			if (!room.settings?.repeats?.length) continue;
+			for (const repeat of room.settings.repeats) {
+				this.runRepeat(room, repeat);
+			}
+		}
+	}
+
+	hasRepeat(room: BasicRoom, phrase: string) {
+		return !!this.repeats.get(room)?.get(phrase);
+	}
+
+	addRepeat(room: BasicRoom, repeat: RepeatedPhrase) {
+		this.runRepeat(room, repeat);
+		if (!room.settings.repeats) room.settings.repeats = [];
+		room.settings.repeats.push(repeat);
+		room.saveSettings();
+	}
+
+	removeRepeat(room: BasicRoom, phrase: string) {
+		if (!room.settings.repeats) return;
+		room.settings.repeats = room.settings.repeats.filter(repeat => repeat.phrase !== phrase);
+		room.saveSettings();
+
+		const roomRepeats = this.repeats.get(room);
+		if (!roomRepeats) return;
+		const oldInterval = roomRepeats.get(phrase);
+		if (oldInterval) clearInterval(oldInterval);
+		roomRepeats.delete(phrase);
+	}
+
+	clearRepeats(room: BasicRoom) {
+		const roomRepeats = this.repeats.get(room);
+		if (!roomRepeats) return;
+		for (const interval of roomRepeats.values()) {
+			if (interval) clearInterval(interval);
+		}
+		this.repeats.delete(room);
+	}
+
+	runRepeat(room: BasicRoom, repeat: RepeatedPhrase) {
+		let roomRepeats = this.repeats.get(room);
+		if (!roomRepeats) {
+			roomRepeats = new Map();
+			this.repeats.set(room, roomRepeats);
+		}
+
+		if (roomRepeats.has(repeat.phrase)) {
+			throw new Error(`Repeat already exists`);
+		}
+
+		roomRepeats.set(
+			repeat.phrase,
+			setInterval(() => {
+				if (room !== Rooms.get(room.roomid)) {
+					// room was deleted
+					this.clearRepeats(room);
+					return;
+				}
+				room.add(`|html|<div class="infobox">${repeat.phrase}</div>`);
+				room.update();
+			}, repeat.interval)
+		);
+	}
+
+	destroy() {
+		for (const roomRepeats of this.repeats.values()) {
+			for (const interval of roomRepeats.values()) {
+				if (interval) clearInterval(interval);
+			}
+		}
+	}
+};
+
+export function destroy() {
+	Repeats.destroy();
+}
+
+export const pages: PageTable = {
+	repeats(args, user) {
+		const room = this.requireRoom();
+		this.title = `Repeated phrases in ${room.roomid}`;
+		this.checkCan("mute", null, room);
+		if (!room.settings.repeats) return `<h1>${this.tr`There are no repeated phrases in ${room.roomid}.`}</h1>`;
+		let html = `<div class="ladder pad"><h1>${this.tr`Repeated phrases in ${room.roomid}`}</h1>`;
```

Comment:
```suggestion
		let html = `<div class="ladder pad"><h2>${this.tr`Repeated phrases in ${room.roomid}`}</h2>`;
```

I'll take the h2, though